### PR TITLE
Feature/snippets

### PR DESCRIPTION
**Pull request by stushurik**

### CODE REVIEW

File: packages/neoFrontend/src/utils/number.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+export const isValidRangeBoundariesString = (rangeBoundariesString) => {
+  const rangeBoundries = rangeBoundariesString.split('-');
+
+  //check if what we split are both numbers, else return nothing
+  return isFiniteNumber(rangeBoundries[0]) && isFiniteNumber(rangeBoundries[1]);
+};
+
+export const parseRangeBoundariesString = (rangeBoundariesString) => {
+  if (isValidRangeBoundariesString(rangeBoundariesString)) {
+    return xorSwap(rangeBoundariesString.split('-'));
```
Comment (curran) [inline packages/neoFrontend/src/utils/number.js:17]: xorSwap is kind of cool, but is it really necessary? Would a simple `.reverse()` do the same thing? Performance is not a concern at all here, just doing this once.

```suggestion
    return rangeBoundariesString.split('-').reverse();
```

File: packages/neoFrontend/src/utils/number.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { range, xorSwap } from './array';
+
+//we check isFinite first since it will weed out most of the non-numbers
+//like mixed numbers and strings, which parseFloat happily accepts
+const isFiniteNumber = (n) => Boolean(isFinite(n) && !isNaN(parseFloat(n)));
```
Comment (curran) [inline packages/neoFrontend/src/utils/number.js:5]: Nice defensive validation.

**Review by curran — APPROVED** [review]
Looks great so far!

I just read through the code. I hope to do a deeper review tomorrow and run it locally.

File: packages/neoFrontend/src/utils/number.js (hunk 1 above)
Comment (stushurik) [reply packages/neoFrontend/src/utils/number.js:17]: xorSwap is smart reverse, if range is 10,20 (asc order) it returns 10,20(stays same), if range is 20,10 (desc order) it returns 10,20 (do reverse)

Comment (stushurik) [reply packages/neoFrontend/src/utils/number.js:17]: I saw it in the example I was inspired by. And at that moment i liked it, but now I think xorSwap is not a good name cause its not obvious what it does

Comment (curran) [reply packages/neoFrontend/src/utils/number.js:17]: Ah I see! That's cool. I'm fine to keep it in that case. It was not clear to me that that's what it does. Maybe a comment would be good to add about that. 
```suggestion
    // xorSwap is smart reverse, if range is 10,20 (asc order) it returns 10,20(stays same), if range is 20,10 (desc order) it returns 10,20 (do reverse)
    return xorSwap(rangeBoundariesString.split('-'));
```

**Review by curran — CHANGES_REQUESTED** [review]
Review comments from QA on https://staging.vizhub.com/curran/7e303867f6764667a45081958825925f?edit=files&file=index.html&line :

 * Linking to line number appears broken now. When clicking a line number, nothing happens.
 * When navigating to the snippet tab, nothing is pre-filled, making it diffucult to get it to work. It would be great to pre-fill all form field values, so the user can start tweaking them instead of trying to figure out what to do to get something to show up. Suggest to default to `index.html` with first 5 lines highlighted.
 * The field "Choose which lines to include into snippet *(requried)" appears to do nothing.
 * The embedded view shows the viewer and top bar.

image

Clicking the "fullscreen" button on the editor and the up arrow gets us pretty close to where we want to be.

image

**Comment by stushurik** [comment]
@curran you saw wrong/broken snippet cause snippet link lead to prod instead of staging.
I pushed a fix, check it again please

**Comment by curran** [comment]
@stushurik I just deployed the latest commit to staging, and the behavior is the same.

More detail: the interaction of clicking on the line number is broken. To reproduce: open the editor, click on a line number.
 * Current behavior with this PR: nothing happens.
 * Expected behavior (in production currently): line number highlights and blinks, and the URL is updated.

**Comment by curran** [comment]
@stushurik This might be a solution for scrolling to center https://stackoverflow.com/questions/10575343/codemirror-is-it-possible-to-scroll-to-a-line-so-that-it-is-in-the-middle-of-w

You can use the average of the start and end to get the line to center on.

File: packages/neoFrontend/src/pages/VizPage/URLStateContext/useURLState.js [hunk 3]
```diff
@@ -74,17 +78,15 @@ export const useURLState = (props) => {
   const vizId = props.match.params.vizId;
   const mode = getMode(state);
 
-  const onHideViz = () => setMode('hide');
-  const onShowViz = () => setMode('viewer');
+  const onHideViz = () => setMode(modes.hide);
```
Comment (curran) [inline packages/neoFrontend/src/pages/VizPage/URLStateContext/useURLState.js:81]: Nice organization.

File: packages/neoFrontend/src/utils/number.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  if (isValidRangeBoundariesString(rangeBoundariesString)) {
+    // xorSwap is smart reverse, if range is 10,20 (asc order) it returns 10,20(stays same), if range is 20,10 (desc order) it returns 10,20 (do reverse)
+    return xorSwap(rangeBoundariesString.split('-'));
+  }
+
+  return null;
+};
+
+
+// inspired by https://codereview.stackexchange.com/questions/26125/getting-all-number-from-a-string-like-this-1-2-5-9
```
Comment (curran) [inline packages/neoFrontend/src/utils/number.js:24]: Nice. Great to include these URLs.

File: packages/neoFrontend/src/utils/number.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+// and http://jsfiddle.net/mkhC3/1/
+export const isNumberSequenceString = (sequenceString) =>
+  sequenceString && !(
+    String(sequenceString)
```
Comment (curran) [inline packages/neoFrontend/src/utils/number.js]: I think it's guaranteed to already be a String, so no need to pass into the `String` constructor?
```suggestion
    sequenceString
```

**Review by curran — APPROVED** [review]
Code changes look great! I put it on staging.

Room for improvement for sure, but this is good enough as a first pass.

I will merge this and create follow-up issues.

In the mean time, we can start testing it out in actual posts and see what the pain points are.

**Pull request merged**